src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -191,18 +191,27 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
             if (previousValue == newValue)
                 return;
 
-            current = new ExecutionContext(current);
-            current.m_localValues[local] = newValue;
-
-            t_currentMaybeNull = current;
+            Dictionary<IAsyncLocal, object> newValues = new Dictionary<IAsyncLocal, object>(current.m_localValues);
+            newValues[local] = newValue;
 
+            List<IAsyncLocal> newChangeNotifications = current.m_localChangeNotifications;
             if (needChangeNotifications)
             {
                 if (hadPreviousValue)
-                    Contract.Assert(current.m_localChangeNotifications.Contains(local));
+                {
+                    Contract.Assert(newChangeNotifications.Contains(local));
+                }
                 else
-                    current.m_localChangeNotifications.Add(local);
+                {
+                    newChangeNotifications = new List<IAsyncLocal>(current.m_localChangeNotifications);
+                    newChangeNotifications.Add(local);
```

Comment:
Rock and hard place...

Creating List with capacity +1; then using InsertRange? InsertRange range also [creates an array](https://github.com/dotnet/coreclr/blob/4cf8a6b082d9bb1789facd996d8265d3908757b2/src/mscorlib/src/System/Collections/Generic/List.cs#L723) which it then copies from so would need to manually iterate and Add, Add, Add?